### PR TITLE
Add bindings for custom un/marshal functions

### PR DESCRIPTION
**Pull request by zekroTJA**
Because large bots need to handle a lot of API requests and responses as well as web socket event messages which all need to go through JSON serialization and deserialization, I thought it might be advantageous if you would be able to swap out the internal JSON implementation with custom ones like [`bytedance/sonic`](https://github.com/bytedance/sonic) as an example.

Therefore, I've added two new types in `types.go` to define the Marshal and Unmarshal functions.
```go
// MarshalFunc encodes an object instance into JSON byte data.
type MarshalFunc func(v interface{}) (data []byte, err error)

// UnmarshalFunc decodes JSON byte data into an object instance.
type UnmarshalFunc func(data []byte, v interface{}) (err error)
```

As with other preferences, I've added them to the `Session` struct.
```go
// The function used to encode object instances
// into JSON data
//
// Defaults to json.Marshal
MarshalFunc MarshalFunc

// The function used to decode JSON data into
// object instances
//
// Defaults to json.Unmarshal
UnmarshalFunc UnmarshalFunc
```

Both values default to the internal `encoding/json` implementation on initialization by default.
```go
func New(args ...interface{}) (s *Session, err error) {
	// Create an empty Session interface.
	s = &Session{
		// ...
		MarshalFunc:            json.Marshal,
		UnmarshalFunc:          json.Unmarshal,
	}
	// ...
```

To utilize the defined functions, It was necessary to pass the `Session` instance to some functions which utilize `Marshal` and `Unmarshal`. Some types which are utilizing these functions as well (like `TimeStamp`), I left untouched to avoid changing the API.

Because I could not get the unit tests to run properly without authentication failures, I've created a quick and dirty example implementation with my fork and sonic as JSON parser. Here you can find the code of it: https://github.com/zekroTJA/discordgo-sonic

---

I've set this repo as `Draft` for the moment just to see if the pull request tests succeed with my changes and make further fixes based on that. Also, I want to implement my fork in one of my own Discord Bots which heavily bases on DiscordGo just to see if there are problems occurring during "field test".

Also, it would be really interesting to get some feedback what you think about this implementation and if you think it fits into the project and makes any sense to implement it this way. :)

Edit: I've now used this implementation in combination with pquerna/ffjson without any issues the last couple of days.

### CODE REVIEW

File: restapi.go [hunk 1]
```diff
@@ -32,7 +31,7 @@ import (
 
 // All error constants
 var (
-	ErrJSONUnmarshal           = errors.New("json unmarshal")
+	ErrJSONUnmarshal           = errors.New("json s.unmarshal")
```
Comment (FedorLap2006) [inline restapi.go:34]: I think you should leave this error as it was, since the meaning didn't change.

File: discord_test.go [hunk 2]
```diff
@@ -23,6 +25,8 @@ var (
 )
 
 func init() {
+	godotenv.Load()
```
Comment (FedorLap2006) [inline discord_test.go:28]: I think would be better to keep out such libraries since (see another comment)

File: go.mod [hunk 3]
```diff
@@ -4,5 +4,6 @@ go 1.13
 
 require (
 	github.com/gorilla/websocket v1.4.2
+	github.com/joho/godotenv v1.3.0
```
Comment (FedorLap2006) [inline go.mod:7]: You add them to go.mod and they become necessary dependency for everyone who uses the library

File: restapi.go [hunk 4]
```diff
@@ -149,9 +148,9 @@ func (s *Session) RequestWithLockedBucket(method, urlStr, contentType string, b
 		}
 	case 429: // TOO MANY REQUESTS - Rate limiting
 		rl := TooManyRequests{}
-		err = json.Unmarshal(response, &rl)
+		err = s.UnmarshalFunc(response, &rl)
 		if err != nil {
-			s.log(LogError, "rate limit unmarshal error, %s", err)
+			s.log(LogError, "rate limit s.unmarshal error, %s", err)
```
Comment (FedorLap2006) [inline restapi.go:153]: Same here.

**Review by FedorLap2006 — CHANGES_REQUESTED** [review]
I think that's a great start, but your code has some problems right now.

File: go.mod (hunk 3 above)
Comment (zekroTJA) [reply go.mod]: Oh yeah, of course. Added it actually just for local testing but yeah, I'm still too accustomed to just do `git add .` sadly. :D

**Comment by zekroTJA** [comment]
> I think that's a great start, but your code has some problems right now.

Thank you very much for your review and feedback. :)

File: go.sum [hunk 5]
```diff
@@ -1,5 +1,7 @@
 github.com/gorilla/websocket v1.4.2 h1:+/TMaTYc4QFitKJxsQ7Yye35DkWvkdLcvGKqM+x0Ufc=
 github.com/gorilla/websocket v1.4.2/go.mod h1:YR8l580nyteQvAITg2hZ9XVh4b55+EU/adAjf1fMHhE=
+github.com/joho/godotenv v1.3.0 h1:Zjp+RcGpHhGlrMbJzXTrZZPrWj+1vfm90La1wgB6Bhc=
+github.com/joho/godotenv v1.3.0/go.mod h1:7hK45KPybAkOC6peb+G5yklZfMxEjkZhHbwpqxOKXbg=
```
Comment (FedorLap2006) [inline go.sum:4]: Not sure why this was added here.

**Review by FedorLap2006 — CHANGES_REQUESTED** [review]

**Comment by FedorLap2006** [comment]
Also note the conflicts.

**Comment by FedorLap2006** [comment]
After a bit of consideration, we've decided that we should do global scoped marshal/unmarshal instead.
The development will be done in the separate PR, so I'm going to close this one.

**Pull request closed without merging**